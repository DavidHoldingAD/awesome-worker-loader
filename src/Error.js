class AwesomeWorkerLoaderError extends Error {
  constructor(message) {
    super(message);

    this.name = 'Awesome Worker Loader Error';
    this.stack = false;
  }
}

export default AwesomeWorkerLoaderError;
