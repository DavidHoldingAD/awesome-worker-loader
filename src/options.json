{
  "type": "object",
  "properties": {
    "name": {
      "type": "string"
    },
    "regExp": {
      "instanceof": "RegExp"
    },
    "inline": {
      "type": "boolean"
    },
    "fallback": {
      "type": "boolean"
    },
    "publicPath": {
      "type": "string"
    },
    "type": {
      "type": "string"
    }
  },
  "additionalProperties": false
}
