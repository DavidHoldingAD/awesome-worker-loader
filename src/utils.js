export function getWorkerConstructorName(workerType) {
  workerType = workerType || 'dedicated';

  const workerTypes = (type) => ({
    'dedicated': 'Worker',
    'shared': 'SharedWorker',
  })[type];

  return workerTypes(workerType);
}
