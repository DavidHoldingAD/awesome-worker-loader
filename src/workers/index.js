/* eslint-disable multiline-ternary */
import path from 'path';

export const WORKER_TYPES = {
  DEDICATED: 'dedicated',
  SHARED: 'shared',
  SERVICE: 'service',
};

const getWorker = (workerFilePath, content, options) => {
  const publicPath = options.publicPath
    ? JSON.stringify(options.publicPath)
    : '__webpack_public_path__';

  const publicWorkerPath = `${publicPath} + ${JSON.stringify(workerFilePath)}`;

  if (options.inline) {
    const fallbackWorkerPath = options.fallback ? publicWorkerPath : 'null';

    const inlineWorkerName = options.name || 'InlineWorker.js';
    const inlineWorkerCreateScript = getInlineWorkerCreateScript(
      inlineWorkerName,
      content,
      fallbackWorkerPath
    );
    return inlineWorkerCreateScript;
  }

  const workerCreateScript = getWorkerCreateScript(
    options.type,
    publicWorkerPath
  );
  return workerCreateScript;
};

const getInlineWorkerCreateScript = (name, content, fallbackWorkerPath) => {
  const inlineWorkerPath = JSON.stringify(`!!${path.join(__dirname, name)}`);

  return `require(${inlineWorkerPath})(${JSON.stringify(
    content
  )}, ${fallbackWorkerPath})`;
};

const getWorkerCreateScript = (type, publicWorkerPath) => {
  switch (type) {
    case WORKER_TYPES.SHARED: {
      return `new SharedWorker(${publicWorkerPath})`;
    }
    case WORKER_TYPES.SERVICE: {
      return `('serviceWorker' in navigator)
      ? navigator.serviceWorker.register(${publicWorkerPath}, options)
      : Promise.reject(
        new Error('navigator.serviceWorker is not supported in this browser')
      );`;
    }
    case WORKER_TYPES.DEDICATED:
    default: {
      return `new Worker(${publicWorkerPath})`;
    }
  }
};

export default getWorker;
