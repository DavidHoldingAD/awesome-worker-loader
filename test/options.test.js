import { pitch } from '../src';
import { WORKER_TYPES } from '../src/workers';

describe('Options', () => {
  [WORKER_TYPES.SHARED, WORKER_TYPES.SERVICE].forEach((type) => {
    test('Options validation', () => {
      const fn = () =>
        pitch.call({
          query: {
            inline: true,
            type,
          },
        });

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });
});
