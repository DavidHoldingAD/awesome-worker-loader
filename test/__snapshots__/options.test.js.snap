// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options Options validation 1`] = `"Cannot inline shared worker."`;

exports[`Options Options validation 2`] = `"Cannot inline service worker."`;
